### PR TITLE
FIX: improves draft for channels

### PR DESCRIPTION
**Pull request by jjaffeux**
This commit attempts to correctly change draft when the channel changes. It moves responsibility to the composer instead of the channel.

A new service `chatDraftsManager` is being introduced here to allow finer control and pave the way for future thread draft support.

These changes also now allow an editing message to be stored as a draft.

### CODE REVIEW

File: plugins/chat/assets/javascripts/discourse/components/chat-channel.js [hunk 1]
```diff
@@ -704,7 +701,7 @@ export default class ChatLivePane extends Component {
           return;
         }
 
-        this.args.channel.draft = null;
+        this.chatDraftsManager.remove({ channelId: this.args.channel.id });
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-channel.js:704]: Both these changes ensure we have removed this draft as soon as the message is sent.

File: plugins/chat/assets/javascripts/discourse/components/chat-composer-message-details.hbs [hunk 2]
```diff
@@ -1,4 +1,8 @@
-<div class="chat-composer-message-details" data-id={{@message.id}}>
+<div
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-composer-message-details.hbs:1]: added to simplify writing specs

File: plugins/chat/assets/javascripts/discourse/components/chat-composer.js [hunk 3]
```diff
@@ -77,15 +78,8 @@ export default class ChatComposer extends Component {
...
-    const message = ChatMessage.createDraftMessage(this.args.channel, {
-      user: this.currentUser,
-      message: raw,
-    });
-
-    this.args.onSendMessage(message);
-
-    return Promise.resolve();
+  sendMessage() {
+    this.args.onSendMessage(this.currentMessage);
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-composer.js:82]: We don't need to create a fake message here, we can't just keep using the draft we have

File: plugins/chat/assets/javascripts/discourse/components/chat-composer.js [hunk 4]
```diff
@@ -106,13 +100,15 @@ export default class ChatComposer extends Component {
 
   @action
   didUpdateMessage() {
-    cancel(this._persistHandler);
+    this.cancelPersistDraft();
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-composer.js:103]: this and next block, we try to cancel current call and persist as soon as possible

File: plugins/chat/assets/javascripts/discourse/components/chat-composer.js [hunk 5]
```diff
@@ -147,11 +143,21 @@ export default class ChatComposer extends Component {
...
+      this.composer.message = null;
+    }
+
+    this.composer.message =
+      this.chatDraftsManager.get({ channelId: this.args.channel.id }) ||
+      ChatMessage.createDraftMessage(this.args.channel, {
+        user: this.currentUser,
+      });
+  }
+
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-composer.js:158]: This is the main change of the PR as we aim to correctly update the composer message each time channel changes

File: plugins/chat/assets/javascripts/discourse/models/chat-message.js [hunk 6]
```diff
@@ -247,6 +247,12 @@ export default class ChatMessage {
       };
     }
 
+    if (this.editing) {
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/models/chat-message.js:250]: allows to now store an editing draft

File: plugins/chat/assets/javascripts/discourse/services/chat-composer.js [hunk 7]
```diff
@@ -1,13 +1,12 @@
...
 import Service, { inject as service } from "@ember/service";
 import { action } from "@ember/object";
-import ChatMessage from "discourse/plugins/chat/discourse/models/chat-message";
 
 export default class ChatComposer extends Service {
   @service chat;
   @service currentUser;
 
-  @tracked _message;
+  @tracked message;
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/services/chat-composer.js:9]: remove uncessary _message

File: plugins/chat/spec/system/chat_composer_spec.rb [hunk 8]
```diff
@@ -5,99 +5,20 @@
   fab!(:channel_1) { Fabricate(:chat_channel) }
   fab!(:message_1) { Fabricate(:chat_message, chat_channel: channel_1) }
 
-  let(:chat) { PageObjects::Pages::Chat.new }
```
Comment (jjaffeux) [inline plugins/chat/spec/system/chat_composer_spec.rb:8]: I move part of this spec into a dedicated chat_composer_draft spec

File: plugins/chat/assets/javascripts/discourse/services/chat-drafts-manager.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import Service from "@ember/service";
+import ChatMessage from "discourse/plugins/chat/discourse/models/chat-message";
+export default class ChatDraftsManager extends Service {
+  drafts = {};
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/services/chat-drafts-manager.js:4]: I wonder if we split this into `threadDrafts` and `channelDrafts`? Or are we just thinking we will only have one draft per channel, including threads?

File: plugins/chat/test/javascripts/unit/services/chat-drafts-manager-test.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    hooks.beforeEach(function () {
+      this.subject = getOwner(this).lookup("service:chat-drafts-manager");
+    });
+
+    hooks.afterEach(function () {
+      this.subject.reset();
+    });
+
+    test("storing and retrivieng messsage", function (assert) {
```
Comment (martin-brennan) [inline plugins/chat/test/javascripts/unit/services/chat-drafts-manager-test.js:19]:
```suggestion
    test("storing and retrieving message", function (assert) {
```

File: plugins/chat/assets/javascripts/discourse/services/chat-drafts-manager.js (hunk 9 above)
Comment (martin-brennan) [reply plugins/chat/assets/javascripts/discourse/services/chat-drafts-manager.js:4]: Then we could have more specific APIs like `addChannelDraft(channelId, message)` etc.

**Review by martin-brennan — APPROVED** [review]
LGTM and works well locally, just a couple of minor comments

**Pull request merged**